pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -119,6 +126,106 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
+// BuildDeltaClusters will generate the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
+// Otherwise, we fall back onto generating everything.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, push *model.PushContext, updates *model.PushRequest) ([]*discovery.Resource, []*discovery.Resource, model.XdsLogDetails) {
+	clusters := make([]*cluster.Cluster, 0)
+	resources := model.Resources{}
+	removedResources := model.Resources{}
+	services := make([]*model.Service, 0)
+	deletedServices := make([]*model.Service, 0)
+	envoyFilterPatches := push.EnvoyFilters(proxy)
+	cb := NewClusterBuilder(proxy, push, configgen.Cache)
+	instances := proxy.ServiceInstances
+	// delta when only services are modified and we attempt deltas
+	// we can progressively optimize this
+	log.Infof("adiprerepa: update types: %v", updates.UpdateTypes)
+	delta := updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.ServiceEntry)
+	// if we can't use delta, fall back to generate all
+	if !delta || updates == nil || len(updates.ConfigsUpdated) == 0 {
+		cl, lg := configgen.BuildClusters(proxy, push)
+		return cl, make([]*discovery.Resource, 0), lg
+	}
+	configs := model.ConfigsOfKind(updates.ConfigsUpdated, gvk.ServiceEntry)
```

Comment:
As all the `ConfigsUpdated` are SE, why need to select again?

Reply:
good catch